Theme preference support on macOS.

weisJ opened the pull request.

weisJ commented:
@vlsi What settings are there on macOS for changing font sizes. Apparently there is only a choice between standard and increased size (only for retina displays). Am I missing something here?

weisJ commented:
I'm having trouble finding how to listen to changes for settings on macOS. For example in the documentation of `[accessibilityDisplayShouldIncreaseContrast](https://developer.apple.com/documentation/appkit/nsworkspace/1526290-accessibilitydisplayshouldincrea)` it says to listen for `accessibilityDisplayOptionsDidChangeNotification` but never states how to do so.
I guess it happens through the `NSDistributedNotificationCenter` but I'm not sure. Haven't found any list of notifications that are broadcasted there.

@vlsi Could you run this snippet to see what values appear? Specifically we're looking for changes to the text scaling, dark mode and high contrast mode.
[[NSDistributedNotificationCenter defaultCenter] addObserver:self selector:@selector(observerMethod:) name:nil object:nil];

vlsi commented:
This seems to be relevant:
https://github.com/ruiaureliano/macOS-Appearance/blob/master/Appearance/Source/AppDelegate.swift#L50
https://medium.com/@ruiaureliano/check-light-dark-appearance-for-macos-mojave-catalina-fb2343af875f
https://stackoverflow.com/a/26472651/1261287

vlsi commented:
> run this snippet to see what values appear

In practice, I need to figure out the syntax because `self` is valid only for Objective-C object

vlsi commented:
```objective-c
@interface Tst:NSObject {
}
@end

@implementation Tst
-(id)init {
    NSDistributedNotificationCenter *center = [NSDistributedNotificationCenter defaultCenter];
    NSLog(@"Adding observer to: %@", center);
    [center addObserver:self
               selector:@selector(notificationEvent:)
                   name:nil
                 object:nil];
   return self;
}

-(void)notificationEvent:(NSNotification *)notif {
    NSLog(@"Notification: %@", notif);
}
@end
```

+

```objective-c
JNIEXPORT void JNICALL
Java_com_github_weisj_darklaf_platform_macos_JNIDecorationsMacOS_setTitleEnabled(JNIEnv *env, jclass obj, jlong hwnd,
                                                                                 jboolean enabled) {
JNF_COCOA_ENTER(env);
    NSLog(@"init");
    Tst *tst = [[Tst alloc]init];
    NSLog(@"init: %@", tst);
```

+ 10.14.6 Mojave

produce the following:

dark-light theme change:
```
2020-04-01 22:13:19.611 java[57577:17801615] Notification: __CFNotification 0x7fb6c532a1b0 {name = ApplePrivateInterfaceThemeChangedNotification}
2020-04-01 22:13:19.611 java[57577:17801615] Notification: __CFNotification 0x7fb6c5010fc0 {name = AppleInterfaceThemeChangedNotification}
```

accent color change:
```
2020-04-01 22:15:35.896 java[57577:17801615] Notification: __CFNotification 0x7fb6c5202b00 {name = AppleAquaColorVariantChanged}
2020-04-01 22:15:35.975 java[57577:17801615] Notification: __CFNotification 0x7fb6c2f4a410 {name = AppleColorPreferencesChangedNotification}
```

switch to "bigger font resolution":

```
2020-04-01 22:17:05.016 java[57577:17801615] Notification: __CFNotification 0x7fb6c8b2b000 {name = QueueStoppedNotification; object = x-coredata://C319EF64-80EB-419D-984C-E29E1BAD56DE/ExchangePrincipal/p2}
2020-04-01 22:17:05.017 java[57577:17801615] Notification: __CFNotification 0x7fb6c8b2dfb0 {name = AppleColorSyncPreferencesSyncNeeded; userInfo = {
    HostScope = kCFPreferencesCurrentHost;
    NotifierPID = 57685;
    UserScope = kCFPreferencesCurrentUser;
}}
2020-04-01 22:17:05.017 java[57577:17801615] Notification: __CFNotification 0x7fb6c2d05190 {name = com.apple.ColorSync.DeviceProfilesNotification; userInfo = {
    DeviceClass = mntr;
    DeviceID = "F466F621-B5FA-04A0-0800-CFA6C258DECD";
    NotifierPID = 57685;
}}
2020-04-01 22:17:05.024 java[57577:17801615] Notification: __CFNotification 0x7fb6c8b433e0 {name = com.apple.ColorSync.DisplayProfileNotification; userInfo = {
    DeviceClass = mntr;
    DeviceID = "F466F621-B5FA-04A0-0800-CFA6C258DECD";
    NotifierPID = 57685;
}}
2020-04-01 22:17:33.120 java[57577:17801615] Notification: __CFNotification 0x7fb6c8a05a80 {name = AppleTextInputMenuExtraDidLoadNotification}
```

activate "inverted screen colors" (in light theme):

```
2020-04-01 22:20:00.376 java[57577:17801615] Notification: __CFNotification 0x7fb6c898a770 {name = com.apple.universalaccess.screenPolarityDidChange; userInfo = {
    pid = 61123;
    uid = 501;
    whiteOnBlack = 1;
}}
```

deactivate "inverted screen colors" (in light theme):

```
2020-04-01 22:21:25.410 java[57577:17801615] Notification: __CFNotification 0x7fb6c8a2c490 {name = com.apple.universalaccess.screenPolarityDidChange; userInfo = {
    pid = 61123;
    uid = 501;
    whiteOnBlack = 0;
}}
```

vlsi commented:
universal access - increase contrast

```
2020-04-01 22:21:52.491 java[57577:17801615] Notification: __CFNotification 0x7fb6c8a2d1d0 {name = AXInterfaceIncreaseContrastStatusDidChange}
```

weisJ commented:
Thank you very much! Do these names correspond to any named notifications. Or at least: Is `accessibilityDisplayShouldIncreaseContrast == "AXInterfaceIncreaseContrastStatusDidChange"`?

vlsi commented:
I guess AXInterfaceIncreaseContrastStatusDidChange corresponds to    https://developer.apple.com/documentation/appkit/nsworkspace/1534227-accessibilitydisplayoptionsdidch

However, I guess the ids can be coded as string literals.

weisJ commented:
> I guess AXInterfaceIncreaseContrastStatusDidChange corresponds to https://developer.apple.com/documentation/appkit/nsworkspace/1534227-accessibilitydisplayoptionsdidch
> 
> However, I guess the ids can be coded as string literals.

Yes that would probably work.
One more thing. Can you check if `[[NSUserDefaults standardUserDefaults]` contains any key-value pair representing the text scaling?

I thin the [medium article](https://medium.com/@ruiaureliano/check-light-dark-appearance-for-macos-mojave-catalina-fb2343af875f) fits the best although I don't quite understand what the comments mean in this snippet:
````
if switchesAutomatically == TRUE
   if appleInterfaceStyle == nil
      theme = dark //will switch to light
   else
      theme = light //will switch to dark
   endif
else
   theme = appleInterfaceStyle //Just like Mojave
endif
```` 
What does `will switch to light/dark` mean? I understand that if `switchesAutomatically == true` the role of `nil` (in `appleInterfaceStyle`) is changed. But what does something that "will" happen have to do with the current theme setting.

vlsi commented:
>What does will switch to light/dark mean? I understand that if switchesAutomatically == true the role of nil (in appleInterfaceStyle) is changed. But what does something that "will" happen have to do with the current theme setting.

Frankly speaking, I've no idea. We should probably find someone with 10.15 :) (I've not updated  yet :) )

>One more thing. Can you check if [[NSUserDefaults standardUserDefaults] contains any key-value pair representing the text scaling?

I do not see UI for adjusting text scaling. Only "resolution".

NSDefaults has the following fonts, however, I don't think they are user-changeable.

```
value: .SFNSText forKey: NSSystemFont
value: 13 forKey: NSSystemFontSize
value: .SFNSText-Bold forKey: NSBoldSystemFont

value: Menlo-Regular forKey: NSFixedPitchFont
value: 11 forKey: NSFixedPitchFontSize

value: Helvetica forKey: NSFont
value: 12 forKey: NSFontSize

value: 4 forKey: AppleAntiAliasingThreshold
```

weisJ commented:
Ok I have implemented the notification listener for dark mode and adjusted the function for checking dark mode.
@vlsi Can you check if thw functions work as espected:
````java
// Does it return the correct value?
JNIThemeInfoMacOS.isDarkThemeEnabled(); 

 // Should be != 0 and a creation message is logged. After creation changes in dark mode should be printed:
long listenerPtr = JNIThemeInfoMacOS.createPreferenceChangeListener();

 // A message should be logged that the listener was destroyed:
JNIThemeInfoMacOS.deletePreferenceChangeListener(listenerPtr);
````

weisJ commented:
> Frankly speaking, I've no idea. We should probably find someone with 10.15 :) (I've not updated yet :) )

Definitely needed 👍 

> I do not see UI for adjusting text scaling. Only "resolution".
 
Too bad. Could it be that if the font size is increased [NSFont systemFontSize] simply returns a larger value?

vlsi commented:
>  Could it be that if the font size is increased [NSFont systemFontSize] simply returns a larger value?

It does not. The values are the same no matter if I increase or decrease "font size".

The font size is adjusted for each monitor individually, so the "font size" is probably reported for each monitor individually as well.

weisJ commented:
What exactly does the `larger text...more space` settings exactly do? Are only cocoa/system apps affected or is it simply scaling the whole screen? Does a swing application look the same regardless of the setting?


Btw. If bigger text ist selected does the titlebar still have the correct size (as the window buttons get scaled too)?

vlsi commented:
> What exactly does the larger text...more space settings exactly do? Are only cocoa/system apps affected or is it simply scaling the whole screen?

It scales everything. Controls, buttons to close Windows, Swing applications, etc, etc.
Icons are scaled as well.

weisJ commented:
Oh so I guess the setting can simply ignored then? As there is no other system setting for font size I’d just always return the default font size on macOS.

As for the other settings:
- inverted screen colours: I guess this is handled by the os already.
- accent color change: Maybe I could add a separate accent colour setting which changes the focus/selection highlights accordingly.

vlsi commented:
> inverted screen colours: I guess this is handled by the os already.

That is indeed handled by OS. I just mean there might be a case for adjusting colors for a better look with inverted colors. I guess "inverting" colors might make contrast worse.

> accent color change: Maybe I could add a separate accent colour setting which changes the focus/selection highlights accordingly.

Well, it can probably be ignored, for now, however, it might be a case for "css-like" themes.
I'm not sure how viable (and/or hard to implement) that is though.

Just a couple of samples:
<img width="780" alt="Снимок экрана 2020-04-02 в 15 32 32" src="https://user-images.githubusercontent.com/213894/78249639-5e88f700-74f7-11ea-86b5-73bc573ad5ab.png">
<img width="780" alt="Снимок экрана 2020-04-02 в 15 32 47" src="https://user-images.githubusercontent.com/213894/78249693-73658a80-74f7-11ea-9498-2939f3a3ffb7.png">
<img width="780" alt="Снимок экрана 2020-04-02 в 15 32 52" src="https://user-images.githubusercontent.com/213894/78249716-7bbdc580-74f7-11ea-8abb-50eb9abc541e.png">
<img width="780" alt="Снимок экрана 2020-04-02 в 15 33 20" src="https://user-images.githubusercontent.com/213894/78249748-85dfc400-74f7-11ea-8d69-3bb967f75b26.png">

weisJ commented:
> > inverted screen colours: I guess this is handled by the os already.
> 
> That is indeed handled by OS. I just mean there might be a case for adjusting colors for a better look with inverted colors. I guess "inverting" colors might make contrast worse.

For now I'll not handle this situation. I guess it's rather specific case.

> 
> > accent color change: Maybe I could add a separate accent colour setting which changes the focus/selection highlights accordingly.
> 
> Well, it can probably be ignored, for now, however, it might be a case for "css-like" themes.
> I'm not sure how viable (and/or hard to implement) that is though.
> 
> Just a couple of samples:
>...

I was thinking of having themes specify two list of colours that are subject to "personalization".
Accent colours and selection colours (the latter is basically just the text selection foreground most of the time).
These colours could then be hue matched with the current system colours. I think this would be the easiest solution that works in a more cross-platform way (i.e. themes don't have to specify all accent/selection values in multiple colours. For example IntelliJ theme would need to specify 19 different values for all macOS colours) that allows basically all available colours (especially because that is the case for text selection on macOS.) I'll try it out and post some screenshots of how it looks with the macOS colours.

weisJ commented:
Ok I have implemented the option to have accent colours mapped.  They can either be completely replaced or give relative values (w.r.t to the new accent colour) in terms of hsb values.

Working with hsb makes it a lot easier to keep the accent hue and only adjust brightness/saturation for hover/click/disabled colours.
Here is how it looks for the IntelliJ theme (the other themes would still need to be prepared for this):
|   |   |
|:-:|:-:|
|![Annotation 2020-04-02 175744](https://user-images.githubusercontent.com/31143295/78271519-88004d80-750c-11ea-9bca-eb17477a0cec.png)|![Annotation 2020-04-01 200843](https://user-images.githubusercontent.com/31143295/78271520-89317a80-750c-11ea-93cc-6002ab0fc843.png)|
|![Annotation 2020-04-02 175713](https://user-images.githubusercontent.com/31143295/78271521-89317a80-750c-11ea-8c18-37d2b8984c42.png)|![Annotation 2020-04-02 175639](https://user-images.githubusercontent.com/31143295/78271524-89ca1100-750c-11ea-956d-afbc5ca089b4.png)|
|![Annotation 2020-04-02 175856](https://user-images.githubusercontent.com/31143295/78271527-89ca1100-750c-11ea-8fea-ef235814e764.png)|![Annotation 2020-04-02 175818](https://user-images.githubusercontent.com/31143295/78271528-8a62a780-750c-11ea-845c-cefe27865710.png)|

WDYT?

weisJ commented:
I guess this could also be done on windows using the accent colour. Though there should probably be a setting to ignore the accent colour (which would also stop it from firing theme preference change events) as there are a lot more accents to choose from some of which might not look that good with some themes (there is no way to guarantee it for arbitrary accents).

In this case a program using darklaf should offer the user to select a custom accent colour or use the default accent colour (provided by the theme). I think that accent colours should also be a feature that is disabled by default and has to be explicitly enabled (I'm not quite certain about how and what part exactly i.e. the system accent isn't taken into consideration or accents are ignored altogether).

High contrast theme should **always** ignore these settings as it doesn't make much sense there.

weisJ commented:
Here are some images for the Darcula theme (I'm using a default button here to show more of the colours):
|   |   |
|:-:|:-:|
|![Annotation 2020-04-02 204945](https://user-images.githubusercontent.com/31143295/78290420-074d4b80-7524-11ea-8e2f-d5d1e97a2328.png)|![Annotation 2020-04-02 205022](https://user-images.githubusercontent.com/31143295/78290423-07e5e200-7524-11ea-9b89-9a44674a5ce2.png)|
|![Annotation 2020-04-02 205103](https://user-images.githubusercontent.com/31143295/78290424-087e7880-7524-11ea-8828-6cabc8e05d34.png)|![Annotation 2020-04-02 205137](https://user-images.githubusercontent.com/31143295/78290425-087e7880-7524-11ea-81aa-073ffbe749f4.png)|
|![Annotation 2020-04-02 205207](https://user-images.githubusercontent.com/31143295/78290426-09170f00-7524-11ea-823e-2b9b841bb65b.png)|![Annotation 2020-04-02 205242](https://user-images.githubusercontent.com/31143295/78290427-09170f00-7524-11ea-8ae4-831c526427db.png)|

weisJ commented:
I have added the support for accent colours. Now it needs to be tested whether the functions work correctly.

kirill-grouchnikov commented:
<img width="532" alt="dark-green1" src="https://user-images.githubusercontent.com/52874/78302453-d4498e80-7508-11ea-943a-4c43e5375f62.png">
<img width="532" alt="light-green1" src="https://user-images.githubusercontent.com/52874/78302454-d4498e80-7508-11ea-92ab-ca64fbe9829b.png">
<img width="607" alt="light-green2" src="https://user-images.githubusercontent.com/52874/78302457-d4e22500-7508-11ea-9b69-0d5779c07d3e.png">
<img width="607" alt="dark-green2" src="https://user-images.githubusercontent.com/52874/78302459-d4e22500-7508-11ea-995c-05d5b98a7b26.png">

These are taken on macOS 10.15.3 with system light/dark and green accent

kirill-grouchnikov commented:
Note that even though the "Continue" button text remains white, the green highlight is different. In light it's around 124,170,83 (not flat color, but close enough) and in dark it's 96,178,64. So in dark it's greener, but darker. In light it's slightly less green and brighter.

weisJ commented:
I have taken the colours from the [Visual Design](https://developer.apple.com/design/human-interface-guidelines/macos/visual-design/color/) page. The fill of checkBoxes should exactly match that value all others might vary. I guess that either
- the colours on website are outdated
- those aren't actually the colours used

Here is how it looks with the colour taken from the screenshot:
![IntelliJ green](https://user-images.githubusercontent.com/31143295/78303340-c281de80-753c-11ea-9215-28aeabab351f.png)

It would really help if I could get the rgb values that are actually used as even though there is an api to get the current accent colour using `[NSColor controlAccentColor]` it is only available in macOS 10.14 and up, which means the library can't be loaded on older versions.
The way I am doing it currently is using `[[NSUserDefaults standardUserDefaults] objectForKey:@"AppleAccentColor"]` which gives an `int` that can be mapped to the current accent colour.

The colours can be obtained by logging `[NSColor controlAccentColor]` with each accent colour selected.

vlsi commented:
I've added `[[NSColor controlAccentColor] colorUsingColorSpace:[NSColorSpace deviceRGBColorSpace]]` and the output is as follows (`[[NSColor controlAccentColor] colorUsingColorSpace:[NSColorSpace sRGBColorSpace]]` produces the same results).

10.14.6

```
accent | r g b
blue   | 0 0.478431 1
lilac  | 0.584314 0.239216 0.588235
rose   | 0.968627 0.309804 0.619608
red    | 0.878431 0.219608 0.243137
orange | 0.968627 0.509804 0.105882
yellow | 0.988235 0.721569 0.152941
green  | 0.384314 0.729412 0.27451
gray   | 0.596078 0.596078 0.596078
```

vlsi commented:
@kirill-grouchnikov , if you hold `alt` during screenshot capture, it would take the screenshot without the shadow

kirill-grouchnikov commented:
Not to derail this conversation, but is there any particular reason to not include shadows - besides a bit of extra size absorbed by Github hosting these images?

vlsi commented:
> but is there any particular reason to not include shadows

The reason here is to fit more images to a single screen, so it is easier to compare how they compare.

kirill-grouchnikov commented:
¯\_(ツ)_/¯

weisJ commented:
> ```
> accent | r g b
> blue   | 0 0.478431 1
> lilac  | 0.584314 0.239216 0.588235
> rose   | 0.968627 0.309804 0.619608
> red    | 0.878431 0.219608 0.243137
> orange | 0.968627 0.509804 0.105882
> yellow | 0.988235 0.721569 0.152941
> green  | 0.384314 0.729412 0.27451
> gray   | 0.596078 0.596078 0.596078
> ```

Are those in dark or light mode? Could you also log in the other mode aswell?

vlsi commented:
> Are those in dark or light mode? Could you also log in the other mode aswell?

The values are the same in both modes.

weisJ commented:
@vlsi @kirill-grouchnikov  Do the functions in `JNIThemeInfoMacOS` return the corrent values for you? It would specifically interesting to know whether `#createPreferenceChangeListener` and `#deletePreferenceChangeListener` produce the console messages `"Listener created"` and `"Listener destroyed"` respectively and if the listener logs a message each time:
- dark mode is toggled
- high contrast mode is toggled
- the accent colours is changed
- the selection colour is changed

weisJ commented:
Ok everything is implemented now. In `core` there is a simple demo `PreferenceChangeDemo`: After pressing the `"Start"` button the theme should automatically adapt to the system settings.

vlsi commented:
Are those `new Thread` really needed?
It looks extremely weird to spawn a thread for LaF "notification wait" purposes :-/

vlsi commented:
What if the listener posts a task via `SwingUtilities.invokeLater(() -> loadnativedefaults());`?
Then the thread won't be required.

vlsi commented:
```java
    protected ThemePreferencesHandler() {
        try {
            // Extend for different platforms.
            boolean enableNativePreferences = isNativePreferencesEnabled();
            if (SystemInfo.isWindows10 && enableNativePreferences) {
                // Decorations are in the Windows10 visuals. Disable for older version.
                preferenceProvider = new WindowsThemePreferenceProvider();
            } else if (SystemInfo.isMac && enableNativePreferences) {
                preferenceProvider = new MacOSThemePreferenceProvider();
            } else {
                preferenceProvider = new DefaultThemePreferenceProvider();
            }
```

^^^ macOS is missing in your PR. For some reason, I can't find it in the diff.

vlsi commented:
After I added MacOSThemePreferenceProvider, it works. Accent color works, light/dark theme switches.

I have not reviewed the code much, however, `new Thread` looks very suspicious :-/

weisJ commented:
> What if the listener posts a task via `SwingUtilities.invokeLater(() -> loadnativedefaults());`?
> Then the thread won't be required.

You would have to get hold of a `JNIEnv` pointer. The one passed to the native functions is only valid inside the thread they come from. You’d have to keep hold of the associated `JVM` pointer and attach the notification thread to the jvm in order to make the call. I’m not quite sure if this would be cheaper because of the attach/detach costs.

An option would be to only attach the thread once and keep it open, but I don’t like the idea of explicitly keeping the native notifications thread attached to the jvm (maybe it’s already attached?).

Effectively the thread is a no-op as it only does something if a change happens. I can change the thread to be a deamon such that it won’t affect jvm closing if something goes wrong.

weisJ commented:
> After I added MacOSThemePreferenceProvider, it works. Accent color works, light/dark theme switches.
> 
> I have not reviewed the code much, however, `new Thread` looks very suspicious :-/

Great! High contrast mode too? Is the listener destroyed when monitoring is stopped?

vlsi commented:
> I’m not quite sure if this would be cheaper because of the attach/detach costs.

The changes to preferences are not frequent, so attach/detach costs do not matter.

> Effectively the thread is a no-op as it only does something if a change happens

The thread would appear in thread dumps, and it would effectively pollute the output.

weisJ commented:
Can you test if the new implementation works @vlsi ? No thread necessary any more.

weisJ commented:
@kirill-grouchnikov I would be very nice if you could confirm that dark mode detection works for you on Catalina.

weisJ commented:
After this is merged I'd like to release a new version, which will probably be version 2.0.0 as it involves some breaking api changes (i.e. themes is separate module, properties have been renamed). I'll use this opportunity to switch to a 3 digit versioning scheme as there is practically no use for the first number currently.

kirill-grouchnikov commented:
Running `PreferenceChangeDemo` from IntelliJ says:

`Error:(34, 15) java: package sun.awt does not exist` in `WindowsTitlePane.java`

Building locally and running `java -cp ./core/build/classes/java/test:./core/build/libs/darklaf-core-1.4.3.1-SNAPSHOT.jar:./macos/build/libs/darklaf-macos-1.4.3.1-SNAPSHOT.jar:./native-utils/build/libs/darklaf-native-utils-1.4.3.1-SNAPSHOT.jar:./platform-base/build/libs/darklaf-platform-base-1.4.3.1-SNAPSHOT.jar:./property-loader/build/libs/darklaf-property-loader-1.4.3.1-SNAPSHOT.jar:./utils/build/libs/darklaf-utils-1.4.3.1-SNAPSHOT.jar:./theme/build/libs/darklaf-theme-1.4.3.1-SNAPSHOT.jar PreferenceChangeDemo` says `java.lang.ClassNotFoundException: com.kitfox.svg.app.beans.SVGIcon`

Creating a separate project doesn't have enough information since it's not clear which Maven repository has the latest SNAPSHOT builds (are they even published)?

I'd suggest making it a bit easier for people to test development / snapshot builds locally on their machines.

weisJ commented:
The aren’t any snapshot builds on maven currently. For now you’ll have to build the project locally to run the tests. I’ll look into releasing some snapshots in the future.

kirill-grouchnikov commented:
Building the project locally in IntelliJ doesn't work because of the references to the `sun.awt` classes.

Running from command line needs a local version of svgSalamander dependency that is not pulled from the original project, but rather from formdev. Eventually I was able to get that and run it as

`java -cp ./core/build/classes/java/test:./core/build/libs/darklaf-core-1.4.3.1-SNAPSHOT.jar:./macos/build/libs/darklaf-macos-1.4.3.1-SNAPSHOT.jar:./native-utils/build/libs/darklaf-native-utils-1.4.3.1-SNAPSHOT.jar:./platform-base/build/libs/darklaf-platform-base-1.4.3.1-SNAPSHOT.jar:./property-loader/build/libs/darklaf-property-loader-1.4.3.1-SNAPSHOT.jar:./utils/build/libs/darklaf-utils-1.4.3.1-SNAPSHOT.jar:./theme/build/libs/darklaf-theme-1.4.3.1-SNAPSHOT.jar:./svgSalamander-1.1.2.1.jar PreferenceChangeDemo`

That just shows an undecorated window with Start / Stop button in the middle with no other UI. 

I don't have additional cycles to spend on figuring out how to run the snapshot demos on my machine. I'd highly recommend creating a list of instructions that can be followed to test changes such as these, especially given the amount of native code that you have for various OSes.

weisJ commented:
Are you building the project using Gradle? The instructions are:
- Press Start.
- change dark mode
- observe if theme changes automatically

kirill-grouchnikov commented:
I did "gradlew build". How do I get from that (which only builds the main libraries, and not the test classes) to running the specific demo? What is the gradle task that runs that demo?

weisJ commented:
There currently are no gradle tasks for starting the demos but you should be able to directly start them from within IntelliJ by clicking the run button in the class.

kirill-grouchnikov commented:
I already said twice that the demos do not start on my Mac machine from IntelliJ due to references to sun.swing classes

vlsi commented:
>I already said twice that the demos do not start on my Mac machine from IntelliJ due to references to sun.swing classes

What is your Java distribution and version?
What is the error? (exact error message)

weisJ commented:
Are you delegating run/compile actions in IntelliJ to gradle?

kirill-grouchnikov commented:
I'm going to step away from this for the day. Once there are reproducible instructions on how to set up the local development environment and how to run the demos, I'll be happy to give that a few minutes.

weisJ commented:
I have added tasks for running the demos.  How to run:

- `git clone https://github.com/weisJ/darklaf.git -bmacos_preferences`
- `cd darklaf`
- `./gradlew PreferenceChangeDemo` 

If you do not have XCode installed (or building the native library fails) you can place [this  binary](https://github.com/weisJ/darklaf/files/4446711/libdarklaf-macos.dylib.zip) in `darklaf/macos/libraries/macos-x86-64/` (You need to unzip it).

kirill-grouchnikov commented:
<img width="538" alt="demo" src="https://user-images.githubusercontent.com/52874/78713320-13e8ef80-78e8-11ea-88e9-910817272d67.png">

This is the only window I see. It stays like this through light/dark switch and accent change at the system level.

vlsi commented:
>It stays like this through light/dark switch and accent change at the system level.

The same for me.

weisJ commented:
Ups. I forgot to add change the monitoring status in the action listener. 😅
Should work now.

kirill-grouchnikov commented:
Tracking light/dark and accent works:
<img width="538" alt="demo1" src="https://user-images.githubusercontent.com/52874/78718661-b1482180-78f0-11ea-91a4-fc14a1189183.png">

Not sure where is the system highlight color is supposed to be seen in the demo. I chose yellow, but the selection is still light blue:
<img width="538" alt="demo2" src="https://user-images.githubusercontent.com/52874/78718660-b0af8b00-78f0-11ea-8deb-e861271b43d2.png">

kirill-grouchnikov commented:
Also, I'd suggest adding a few controls here that always show the accent color (checkbox? combobox?) so that you can immediately see the accent color taking place across the UI (without having to transfer focus to that button up top).

kirill-grouchnikov commented:
One accent that doesn't work (under light and dark) is the default Aqua / blue. When that is set at the system level, the demo button shows strong red outline:

<img width="538" alt="demo3" src="https://user-images.githubusercontent.com/52874/78718949-20257a80-78f1-11ea-95ec-b40f5c373503.png">

weisJ commented:
> Tracking light/dark and accent works:

👍 How about high contrast mode?

> Not sure where is the system highlight color is supposed to be seen in the demo. I chose yellow, but the selection is still light blue:

Not intended. I have changed `selectedControlColor` to `selectedControlTextColor` maybe that works.

>Also, I'd suggest adding a few controls here that always show the accent color (checkbox? >combobox?) so that you can immediately see the accent color taking place across the UI (without >having to transfer focus to that button up top).

I have added a solid square to display the colours.

>One accent that doesn't work (under light and dark) is the default Aqua / blue. When that is set at >the system level, the demo button shows strong red outline:

Should be fixed now.

kirill-grouchnikov commented:
Which one of these settings are you treating as high contrast mode?

<img width="780" alt="accessibility" src="https://user-images.githubusercontent.com/52874/78721143-f79f7f80-78f4-11ea-9055-02aa06519193.png">

weisJ commented:
Increase Contrast

kirill-grouchnikov commented:
Selection color always is shown as black, no matter what is the system selection.

Also, the first accent color switch is not reflected in the demo. Second and all consecutive switches are reflected correctly.

kirill-grouchnikov commented:
I've never used this setting, so I don't know if the demo look is correct

<img width="533" alt="contrast1" src="https://user-images.githubusercontent.com/52874/78721992-924c8e00-78f6-11ea-8c76-717af54d3647.png">
<img width="533" alt="contrast2" src="https://user-images.githubusercontent.com/52874/78721990-91b3f780-78f6-11ea-9829-94ddcea587e9.png">

weisJ commented:
High contrast mode looks good.  Thank you very much for the testing  👍 
Does the first change only affect the accent colour or everything? Does it still occur when switching dark mode first?
As for the selection colour I’ll have to look further into it.

kirill-grouchnikov commented:
Dark/light selection gets reflected every time, straight away.
It's the accent color selection that is not "catching" the very first system-level change.

For example:

1. Open app and click Start
2. Open system preferences
3. Change to dark mode - the demo is updated
4. Change accent selection - the demo is not updated
5. Change accent selection - the demo is updated

vlsi commented:
@weisJ , just in case, "selection color" is always black since:

> selectedControlTextColor
>The color to use for text in a selected control—that is, a control being clicked or dragged.

The selection background is `selectedTextBackgroundColor`

weisJ commented:
I have added some logging statements for when a native notification is received.
Does switching the accent colour the first time trigger a notification? If yes what is logged.

vlsi commented:
Now "selection color" indicator is "late by one".
In other words, it does not react to the very first change, then it displays "previous" selection color.

TC:
0. Light, default blue accent.
1. Open demo & start
2. Change selection color to rose.
AR: nothing happens
log:

```
> Task :PreferenceChangeDemo
[2020-04-08 11:37:59] [INFO] Loaded libdarklaf-macos.dylib. Native features are enabled. [at com.github.weisj.darklaf.platform.macos.MacOSLibrary]
[2020-04-08 11:38:07] [INFO] Started preference monitoring. [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:38:11] [INFO] Event received: [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:38:11] [INFO]      Dark Mode:       [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:38:11] [INFO]      High Contrast:   [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:38:11] [INFO]      Accent Color:    [old: java.awt.Color[r=0,g=122,b=255] new: java.awt.Color[r=0,g=122,b=255]] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:38:11] [INFO]      Selection Color: [old: java.awt.Color[r=164,g=204,b=254] new: java.awt.Color[r=164,g=204,b=254]] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
```

weisJ commented:
What happens the second time the selection colour is changed?

vlsi commented:
The icon in the demo **always** shows the color that was selected **before** the change.

1) Starting position: blue (setting) | blue (demo)
2) Updated to rose: rose (setting) | blue (demo)
log:
```
[2020-04-08 11:53:32] [INFO] Event received: [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:53:32] [INFO]      Dark Mode:       [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:53:32] [INFO]      High Contrast:   [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:53:32] [INFO]      Accent Color:    [old: java.awt.Color[r=0,g=122,b=255] new: java.awt.Color[r=0,g=122,b=255]] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:53:32] [INFO]      Selection Color: [old: java.awt.Color[r=164,g=204,b=254] new: java.awt.Color[r=164,g=204,b=254]] [at com.github.weisj.darklaf.platform.macos.M
```

3) Update to yellow: yellow (setting) | rose (demo)

```
[2020-04-08 11:54:42] [INFO] Event received: [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:54:42] [INFO]      Dark Mode:       [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:54:42] [INFO]      High Contrast:   [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:54:42] [INFO]      Accent Color:    [old: java.awt.Color[r=0,g=122,b=255] new: java.awt.Color[r=0,g=122,b=255]] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:54:42] [INFO]      Selection Color: [old: java.awt.Color[r=164,g=204,b=254] new: java.awt.Color[r=249,g=187,b=219]] [at com.github.weisj.darklaf.platform.macos.M
```

4) update to green: green (setting) | yellow (demo)

```
[2020-04-08 11:55:21] [INFO] Event received: [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:55:21] [INFO]      Dark Mode:       [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:55:21] [INFO]      High Contrast:   [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:55:21] [INFO]      Accent Color:    [old: java.awt.Color[r=0,g=122,b=255] new: java.awt.Color[r=0,g=122,b=255]] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:55:21] [INFO]      Selection Color: [old: java.awt.Color[r=249,g=187,b=219] new: java.awt.Color[r=253,g=228,b=176]] [at com.github.weisj.darklaf.platform.macos.M
```

5) update to green: green(setting) | green(setting)

```
[2020-04-08 11:56:06] [INFO] Event received: [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:56:06] [INFO]      Dark Mode:       [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:56:06] [INFO]      High Contrast:   [old: false new: false] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:56:06] [INFO]      Accent Color:    [old: java.awt.Color[r=0,g=122,b=255] new: java.awt.Color[r=0,g=122,b=255]] [at com.github.weisj.darklaf.platform.macos.MacOSPreferenceMonitor]
[2020-04-08 11:56:06] [INFO]      Selection Color: [old: java.awt.Color[r=253,g=228,b=176] new: java.awt.Color[r=198,g=230,b=186]] [at com.github.weisj.darklaf.platform.macos.M
```

weisJ commented:
Oh ok. The question now is: Does the laf change or is it only the icon that has the wrong color?

weisJ commented:
I'm not running the callback on the cocoa main thread. I hope this ensures that all values are correctly set.

weisJ commented:
@vlsi is the selection colour now in sync or still one change behind?

vlsi commented:
1) The icon color is still "delayed by one"
2) The selection color itself is not changed
3) Probably you need to track both `selectedControlTextColor` and `selectedTextBackgroundColor`

weisJ commented:
> 2. The selection color itself is not changed

Do you mean the selection colour of textfields isn't changed at all?

> 3. Probably you need to track both `selectedControlTextColor` and `selectedTextBackgroundColor`

The change event is received but `selectedTextBackgroundColor` returns the old value instead of the new one. There is no difference in tracking one or both of them as both are reported through `NSSystemColorsDidChangeNotification`.

weisJ commented:
If you log `selectedTextBackgroundColor` directly in the event listener does it return the correct value for you?

vlsi commented:
>Do you mean the selection colour of textfields isn't changed at all?

That is true

weisJ commented:
> > Do you mean the selection colour of textfields isn't changed at all?
> 
> That is true

Does 
````java
LafManager.setTheme(new PreferredThemeStyle(ContrastRule.STANDARD,
                                            ColorToneRule.LIGHT,
                                            AccentColorRule.fromColor(MacOSColors.BLUE, Color.RED)));
````
produce a red selection colour?

weisJ commented:
I'm now retrieving the selection colour using `NSColor *accentColor = [[NSColorList colorListNamed: @"System"] colorWithKey:@"selectedTextBackgroundColor"];`
Maybe this value is updated earlier.

vlsi commented on macos/src/main/objectiveCpp/ThemeInfo.mm:
```diff
@@ -0,0 +1,193 @@
+JNIEXPORT jint JNICALL
+Java_com_github_weisj_darklaf_platform_macos_JNIThemeInfoMacOS_nativeGetSelectionColor(JNIEnv *env, jclass obj) {
+JNF_COCOA_ENTER(env);
+    NSColorSpace *rgbSpace = [NSColorSpace genericRGBColorSpace];
+    NSColor *accentColor = [[NSColorList colorListNamed: KEY_SYSTEM_COLOR_LIST] colorWithKey:KEY_SELECTION_COLOR];
```
```
Non-Java exception raised, not handled! (Original problem: *** -redComponent not valid for the NSColor Catalog color: System selectedTextBackgroundColor; need to first convert colorspace.)
        at com.github.weisj.darklaf.platform.macos.JNIThemeInfoMacOS.nativeGetSelectionColor(Native Method)
```

Need to convert the color to the colorspace.

weisJ replied:
Ups. Fixed

vlsi replied:
```
-[NSDynamicSystemColor usingColorSpace:]: unrecognized selector sent to instance 0x7fa0fad70fa0
Exception in thread "AWT-EventQueue-0" java.lang.RuntimeException: Non-Java exception raised, not handled! (Original problem: -[NSDynamicSystemColor usingColorSpace:]: unrecognized selector sent to instance 0x7fa0fad70fa0)
        at com.github.weisj.darklaf.platform.macos.JNIThemeInfoMacOS.nativeGetSelectionColor(Native Method)
```

vlsi commented:
OK, now the behavior is:
1) If accent color is changed (it changes selection automatically), then PreferenceChangeDemo is updated
2) If only selection color is changed, then PreferenceChangeDemo is not updated. However, if PreferenceChangeDemo is activated, then it is re-painted with the appropriate color

In all the cases, the selection is old:
<img width="760" alt="selection color" src="https://user-images.githubusercontent.com/213894/78988112-a5e73700-7b38-11ea-9bc9-babb4c0ea4aa.png">

weisJ commented:
Apparently `[self listenToKey:@"AppleColorPreferencesChangedNotification" onCenter:center];` does only fire of the system menu looses focus. But `[self listenToKey:nil onCenter:center];` does fire even if the the focus is not switched. I could simply listen to everything and filter the messages or simply leave it as is (as selections are shown if the window has focus anyway).

weisJ commented:
The selection colour should now actually change
@vlsi What is your opinion on my previous comment?

weisJ commented:
Looks like the check for `AppleInterfaceStyleSwitchesAutomatically` isn't actually necessary. See [here](https://github.com/weisJ/auto-dark-mode/issues/1)

weisJ commented:
If there are no more objections I’d merge this PR.
